[Boost]: Add settings system to Boost Cache.

donnchawp opened the pull request:
This PR adds a settings system to Boost Cache which is shared between the auto loaded Page_Cache Module and the caching system loaded by advanced-cache.php.

Fixes [#35216](https://github.com/Automattic/jetpack/issues/35216)
## Proposed changes:
<!--- Explain what functional changes your PR includes -->
* This adds the Boost_Class_Settings class which includes the get() and set() methods.
* It will have an internal $settings arrray.
* get() will return the value of one settings at a time.
* set() accepts an associative array of settings to save to the config file. The internal $settings array is updated at the same time.
* The config file is stored in wp-content/boost-cache/ and is named config.php
* When run for the first time, it will create a new config file with the setting `enabled => true`.
* The config file is a php script and the first line is <?php die(); The second line is simply a JSON encoded array of the settings.
* The config file is loaded using the file() command.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
pc9hqz-2rA-p2

## Does this pull request change what data or activity we track or use?
No

## Testing instructions:
* Apply PR
* Reload your blog.
* Check for the new config file at wp-content/boost-cache/config.php

donnchawp committed:
Add settings class to store cache configuration.

donnchawp committed:
changelog

dilirity reviewed (changes requested):
If there's no `boost-cache` directory, the config is not created even though boost cache is enabled.

I'm testing this by deleting the `boost-cache` directory and refreshing the Boost admin page. I'm not opening the front-end intentionally, so the folder doesn't get created.

donnchawp committed:
Create boost-cache directory for settings file

donnchawp commented:
Good catch @dilirity. Just fixed that.

dilirity commented on projects/plugins/boost/app/modules/cache/Boost_Cache_Settings.php:
```diff
@@ -0,0 +1,106 @@
+
+		$this->settings = array_merge( $this->settings, $settings );
+
+		$contents     = "<?php die();\n" . json_encode( $this->settings ); // phpcs:ignore WordPress.WP.AlternativeFunctions.json_encode_json_encode
```
This shows an error when accessing the file. Adding an opening multiline comment on a new line should fix this.

![CleanShot 2024-01-29 at 14 09 03](https://github.com/Automattic/jetpack/assets/11799079/2504ba83-7224-49a0-80b0-b433dd443f18)

On further thought, maybe we should include a comment here, saying that this is the stored cache settings. It might be mistaken for hacky code.

donnchawp replied:
You're right. A multiline comment would solve it, and give us room to explain what it is.

dilirity commented on projects/plugins/boost/app/modules/cache/Boost_Cache_Settings.php:
```diff
@@ -0,0 +1,106 @@
+
+		if ( ! file_exists( $this->config_file ) ) {
+			$result = $this->set( array( 'enabled' => true ) );
+			// TODO: handle $result if it's false
```
Is this a left over or is it something that needs doing?

donnchawp replied:
Oh yeah, I meant to take that out.

donnchawp replied:
I left it in, but slightly modified it and made "enabled" = false. We don't check that "enabeld" settings yet so it does nothing right now.

dilirity reviewed (changes requested):
Two small things and I think it's ready for merge.

I've been wondering about last_error - should we maybe add that to the `debug.log`? What do you think?

donnchawp committed:
Make config file friendlier

donnchawp committed:
Disable caching by default when creating a new config file

thingalon commented on projects/plugins/boost/app/modules/cache/Boost_Cache.php:
```diff
@@ -6,8 +6,14 @@
+require_once __DIR__ . '/Boost_Cache_Settings.php';
 
 abstract class Boost_Cache {
+	/*
+	 * @var array - The settings for the page cache.
+	 */
+	private $settings = array();
```
The comment and initialization says this is an array, but it gets assigned a new `Boost_Cache_Settings` object in the constructor.

donnchawp replied:
Good catch. It's an array in the settings class.

thingalon replied:
Still says `array` in the docblock.

thingalon commented on projects/plugins/boost/app/modules/cache/Boost_Cache_Settings.php:
```diff
@@ -0,0 +1,107 @@
+			$this->last_error = 'Invalid config file';
+			return false;
+		}
+		$settings = json_decode( $lines[3], true );
```
This feels fragile; a human editing this file (and I promise that some will try) will easily break this. Can I suggest that it reads the whole file into a string, finds the first `{` and decodes everything from that character onwards? 

That way if users add comments, extra newlines, or whatever, this will survive it.

donnchawp replied:
Good idea. I've done that, and added a warning, "Do not edit." too.
If they really mess up, they can delete the file.

thingalon commented on projects/plugins/boost/app/modules/cache/Boost_Cache_Settings.php:
```diff
@@ -0,0 +1,107 @@
+	 * @param string $setting - The setting to get.
+	 * @return mixed - The value of the setting, or false if the setting does not exist. Call get_last_error() to get the error message.
+	 */
+	public function get( $setting ) {
+		if ( ! isset( $this->settings[ $setting ] ) ) {
+			$this->last_error = 'Setting not found';
+			return false;
+		}
+		return $this->settings[ $setting ];
+	}
```
🤔 Thinking out loud here: what do we generally want it to do when a specific setting isn't found?

I think it should generally go with a default. e.g.: if "enabled" is missing, go with `false`. If `exclusions` is missing, go with `array()`.

Perhaps let `get` take a second parameter - a default value to return if the setting is missing. Then add functions like `get_enabled()` which will return `true` or `false` for enabled, and `get_exclusions` which will always return an array of strings?

That way, the defaults are all defined in this settings file (rather than with every bit of code that asks for a setting), and all the code that asks for settings can just ask for the setting it wants.

What do you think?

donnchawp replied:
Great idea. Thank you!

thingalon commented on projects/plugins/boost/app/modules/cache/Boost_Cache_Settings.php:
```diff
@@ -0,0 +1,107 @@
+
+		$this->settings = array_merge( $this->settings, $settings );
+
+		$contents     = "<?php die();\n/*\n * Configuration data for Jetpack Boost Cache.\n" . json_encode( $this->settings ) . "\n */"; // phpcs:ignore WordPress.WP.AlternativeFunctions.json_encode_json_encode
+		$tmp_filename = $this->config_file . uniqid( uniqid(), true ) . '.tmp';
+		if ( false === file_put_contents( $tmp_filename, $contents ) ) { // phpcs:ignore WordPress.WP.AlternativeFunctions.file_system_operations_file_put_contents
+			$this->last_error = 'Could not write to tmp config file';
+			return false;
+		}
```
We're now doing this write-to-tmp and rename trick twice. 

I think it's time for this behaviour to go in a utility function.

donnchawp replied:
Agreed. The Utils class is in the deactivate PR, so once both of these are merged I'll do that. Issue added in [#35331](https://github.com/Automattic/jetpack/issues/35331)

thingalon commented on projects/plugins/boost/app/modules/cache/Boost_Cache_Settings.php:
```diff
@@ -0,0 +1,107 @@
+		}
+
+		if ( ! file_exists( $this->config_file ) ) {
+			$result = $this->set( array( 'enabled' => false ) );
+			if ( ! $result ) {
+				return false;
+			}
```
At this point, if `$result` is true, don't we already know the value of settings? Is there a reason to read it back?

donnchawp replied:
$result is true only if the setting was successfully written to the config file. We're not doing anything with that return value yet anyway, so I'll remove it. It would be better off handled in the set() function, where it will log it for an admin_notice perhaps to inform the user.

thingalon replied:
Wait, if you remove that check then it will try to read the config file after it fails to save it and overwrite the last_error message about the save failing.

My point was that if $result is true, you can just return the value you wrote. :) But I guess that's only a tiny optimization for the smallest edge-case.  And the last_error thing doesn't super matter until we implement error logging.

thingalon commented on projects/plugins/boost/app/modules/cache/Page_Cache.php:
```diff
@@ -5,7 +5,18 @@
+require_once __DIR__ . '/Boost_Cache_Settings.php';
+
 class Page_Cache implements Pluggable, Is_Always_On {
+	/*
+	 * @var array - The settings for the page cache.
+	 */
+	private $settings = array();
```
Not an array.

donnchawp replied:
Thanks, fixed that.

thingalon replied:
Comment still says it's an array.

thingalon commented on projects/plugins/boost/app/modules/cache/Page_Cache.php:
```diff
@@ -5,7 +5,18 @@
+	 */
+	private $settings = array();
+
+	public function __construct() {
+		$this->settings = new Boost_Cache_Settings();
+	}
```
So, `Page_Cache` is used when visiting admin pages, and `Boost_Cache` is used when actually driving the caching process. Will they ever co-exist? Is it worth making Settings a singleton, or will only one exist at a time?

donnchawp replied:
When a wp-admin request is served, the `Boost_Cache` class is loaded by advanced-cache.php, and it creates an instance of Boost_Cache_Settings, but it does nothing because `$this->is_cacheable()` in `get()` returns false, and the same in ob_start().

However, this does mean it went and read the config file, which is a bit of a waste. I also noticed a bug in the is_backend() function - it checks $script else checks "REST_REQUEST", when the "REST_REQUEST" bit won't ever run because $script is "index.php".

So, I'll fix that bug and add an is_backend() check in the Boost_Cache constructor.

donnchawp replied:
I just realised that the code that defines REST_REQUEST runs after Boost_Cache has been created, so for those requests, the output buffer is created. It's not saved in the callback though, as it recognises it's a REST request.

The Boost_Cache constructor could check if the REQUEST_URI is /wp-admin/ or /wp-json/ but that only helps if the site is at "/". A page URL might have either of those strings in it.

So, the settings object is created twice in wp-admin and wp-json requests. I think it might be worth making the settings class a singleton.

thingalon reviewed (changes requested):
Got some thoughts and suggestions.

donnchawp committed:
Settings here is not an array.

donnchawp committed:
Look for "{" to decode the JSON in the config file.

donnchawp committed:
Warn curious minds they should not edit this file.

donnchawp committed:
Move these checks out of $script check to simplify it.

donnchawp committed:
Settings aren't an array here.

donnchawp committed:
Convert settings class to a singleton, as a shared resource.

donnchawp committed:
Just set this setting, no need to check return value.

Any problems should be handled in the set() function and communicated to
the user there.

donnchawp committed:
Add $default setting to get(), and helper get functions

Commit:
Merge branch 'trunk' into boost-cache/settings_class

thingalon reviewed (approved):
Everything left seems minor, so I'm going to ✅ and suggest that the small stuff gets dealt with in a small standalone PR after.
